### PR TITLE
loopout: user-specified expiry

### PR DESCRIPTION
**Pull request by joostjager**
This PR allows the user to select the cltv delta of the on-chain htlc, within boundaries set by the client. The selection takes place implicitly through the sweep confirmation target swap parameter.

The range of accepted cltv delta is observable through the `terms` command.

Related: #229 #207

### CODE REVIEW

File: looprpc/server.proto [hunk 1]
```diff
@@ -177,3 +206,76 @@ message ServerLoopOutPushPreimageRequest {
 
 message ServerLoopOutPushPreimageResponse {
 }
+
+message SubscribeUpdatesRequest {
```
Comment (carlaKC) [inline looprpc/server.proto:210]: Shouldn't these rather be added in the subscribe PR since they're not used?

File: looprpc/server.proto [hunk 2]
```diff
@@ -177,3 +206,76 @@ message ServerLoopOutPushPreimageRequest {
...
+    FAILED_SWAP_DEADLINE = 8;
+
+    // The server could not publish the loop out on chain htlc.
+    FAILED_HTLC_PUBLICATION = 9;
+
+    // The server has published the loop out on chain timeout tx.
+    TIMEOUT_PUBLISHED = 10;
+
+    // The swap has failed for unknown reasons, it will not be completed.
+    UNEXPECTED_FAILURE = 11;
```
Comment (carlaKC) [inline looprpc/server.proto:264]: If we are keeping these, need the htlc confirmed state that was just added.

File: client.go [hunk 3]
```diff
@@ -386,6 +390,26 @@ func (s *Client) LoopOut(globalCtx context.Context,
...
+// getExpiry returns an absolute expiry height based on the sweep confirmation
+// target, constrained by the server terms.
+func (s *Client) getExpiry(height int32, terms *LoopOutTerms,
+	confTarget int32) (int32, error) {
+
+	cltvDelta := confTarget
+
+	switch {
+	case confTarget < terms.MinCltvDelta:
+		cltvDelta = terms.MinCltvDelta
```
Comment (carlaKC) [inline client.go:402]: could just return inline here then we don't need cltvDelta

Comment (bhandras) [reply client.go:402]: won't the linter complain?

Comment (carlaKC) [reply client.go:402]: looks ok?
```
	switch {
	case confTarget < terms.MinCltvDelta:
		return height + terms.MinCltvDelta, nil

	case confTarget > terms.MaxCltvDelta:
		return 0, fmt.Errorf("...")
	}

	return height + confTarget, nil
```

Comment (bhandras) [reply client.go:402]: oh yes, true!

File: client.go [hunk 4]
```diff
@@ -386,6 +390,26 @@ func (s *Client) LoopOut(globalCtx context.Context,
...
+
+	case confTarget > terms.MaxCltvDelta:
+		return 0, fmt.Errorf("confirmation target %v exceeds maximum "+
```
Comment (bhandras) [inline client.go:405]: Alternatively we could just clamp to max just like in the min case?

File: looprpc/server.proto (hunk 1 above)
Comment (bhandras) [reply looprpc/server.proto:210]: +1

File: client.go (hunk 4 above)
Comment (joostjager) [reply client.go:405]: That is what I first had, but realized that we also need to protect users against sweeping with very high conf targets while the expiry is much sooner

File: looprpc/server.proto (hunk 2 above)
Comment (joostjager) [reply looprpc/server.proto:264]: added

File: client.go (hunk 3 above)
Comment (joostjager) [reply client.go:402]: fixed

File: client.go [hunk 5]
```diff
@@ -386,6 +390,24 @@ func (s *Client) LoopOut(globalCtx context.Context,
...
 }
 
+// getExpiry returns an absolute expiry height based on the sweep confirmation
+// target, constrained by the server terms.
+func (s *Client) getExpiry(height int32, terms *LoopOutTerms,
+	confTarget int32) (int32, error) {
+
+	switch {
+	case confTarget < terms.MinCltvDelta:
+		return terms.MinCltvDelta, nil
```
Comment (carlaKC) [inline client.go:400]: height + MinCltvDelta?

**Comment by carlaKC** [comment]
This is pretty much there for me, just need to fix the unit tests. Still uncertain about adding protos that aren't used in the PR though.

File: client.go (hunk 5 above)
Comment (joostjager) [reply client.go:400]: Ah, that is why I didn't do the early retuns I remember now.

Comment (joostjager) [reply client.go]: Fixed

**Pull request merged**